sa/sa.go

```diff
@@ -950,47 +950,51 @@ func (ssa *SQLStorageAuthority) leaseOldestCRLShard(ctx context.Context, req *sa
 			&shards,
 			`SELECT id, issuerID, idx, thisUpdate, nextUpdate, leasedUntil
 				FROM crlShards
-				WHERE issuerID = ?`,
-			req.IssuerNameID,
+				WHERE issuerID = ?
+				AND idx BETWEEN ? AND ?`,
+			req.IssuerNameID, req.MinShardIdx, req.MaxShardIdx,
 		)
 		if err != nil {
 			return -1, fmt.Errorf("selecting candidate shards: %w", err)
 		}
 
-		// Convert the slice to a map to detect shards that we expect to exist, but
-		// don't. At the same time, determine the oldest known shard.
-		shardMap := make(map[int]*crlShardModel, len(shards))
-		var oldest *crlShardModel
-		for _, shard := range shards {
-			if shard.Idx < int(req.MinShardIdx) || shard.Idx > int(req.MaxShardIdx) {
-				continue
+		// Determine which shard index we want to lease.
+		var shardIdx int
+		var needToInsert bool
+		if len(shards) < (int(req.MaxShardIdx + 1 - req.MinShardIdx)) {
```

Comment:
What is the purpose of this `+ 1` here and line 967, but then only ranging between `req.MinShardIdx` and `req.MaxShardIdx` below on line 968.

Reply:
Note that req.MaxShardIdx is an *inclusive* upper bound here. So this line and line 968 are in sync: this one adds one to the difference, to account for including both endpoints of the range; that line uses `<=` instead of `<` to include both endpoints of the range.

I could have phrased line 968 as `i < req.MaxShardIdx + 1` I suppose; would that be more readable?

Reply:
I think it's fine to leave it as-is. Thanks for holding my hand through this. 😫